File: companion/app/src/main/java/com/example/sentrykey/CloudCrypto.kt
package com.example.sentrykey

import android.util.Base64
import org.json.JSONObject
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Zero-knowledge cloud-account crypto. Byte-for-byte compatible with the web
 * dashboard's `crypto.js` (deriveUserKeys + encryptWithKey/decryptWithKey), so a
 * vault backed up from the phone can be opened in the browser and vice-versa.
 *
 *   masterKey = PBKDF2-HMAC-SHA256(password, salt = lowercased username, 210k, 256-bit)
 *   authKey   = HMAC-SHA256(masterKey, "auth-key")        -> hex, sent to the server
 *   encKey    = HMAC-SHA256(masterKey, "encryption-key")  -> 32 bytes, NEVER leaves device
 *
 * The vault is sealed with AES-256-GCM using encKey directly (the envelope salt
 * is decorative for this path, matching the web). The server only ever stores
 * the encrypted envelope + the authKey hash; it cannot read secrets.
 */
object CloudCrypto {
    private const val ITERATIONS = 210_000
    private const val KEY_BITS = 256
    private const val SALT_LEN = 16
    private const val IV_LEN = 12
    private const val TAG_BITS = 128

    data class UserKeys(val authKey: String, val encKey: ByteArray)

    /** Derives the server auth key (hex) and the local vault encryption key. */
    fun deriveUserKeys(username: String, password: String): UserKeys {
        val cleanUsername = username.trim().lowercase()
        val spec = PBEKeySpec(password.toCharArray(), cleanUsername.toByteArray(Charsets.UTF_8), ITERATIONS, KEY_BITS)
        val masterKey = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded

        val authKey = hmacSha256(masterKey, "auth-key").toHex()
        val encKey = hmacSha256(masterKey, "encryption-key")
        return UserKeys(authKey, encKey)
    }

    /** Encrypts [plaintext] with a pre-derived [encKey] into a SentryKey .skbackup envelope. */
    fun encryptWithKey(plaintext: String, encKey: ByteArray): String {
        val random = SecureRandom()
        val salt = ByteArray(SALT_LEN).also { random.nextBytes(it) }
        val iv = ByteArray(IV_LEN).also { random.nextBytes(it) }

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(encKey, "AES"), GCMParameterSpec(TAG_BITS, iv))
        val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

        return JSONObject().apply {
            put("app", "SentryKey")
            put("version", 1)
            put("encrypted", true)
            put("kdf", "PBKDF2WithHmacSHA256")
            put("iterations", ITERATIONS)
            put("salt", b64(salt))
            put("iv", b64(iv))
            put("ciphertext", b64(ciphertext))
        }.toString(2)
    }

    /** Reverses [encryptWithKey]. Throws [BadPasswordException] if the key is wrong. */
    fun decryptWithKey(envelopeJson: String, encKey: ByteArray): String {
        val obj = try {
            JSONObject(envelopeJson)
        } catch (e: Exception) {
            throw BadPasswordException("Not a valid SentryKey backup.")
        }
        if (!obj.optBoolean("encrypted", false)) {
            throw BadPasswordException("Backup is not encrypted.")
        }
        try {
            val iv = unb64(obj.getString("iv"))
            val ciphertext = unb64(obj.getString("ciphertext"))
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(encKey, "AES"), GCMParameterSpec(TAG_BITS, iv))
            return String(cipher.doFinal(ciphertext), Charsets.UTF_8)
        } catch (e: Exception) {
            throw BadPasswordException("Wrong account password or corrupt backup.")
        }
    }

    // ---- Account recovery (matches web crypto.js generateRecoveryKey/deriveRecovery/wrap) ----

    data class RecoveryKeys(val wrapKey: ByteArray, val authKey: String)

    /** A readable one-time recovery key, e.g. ABCDE-FGHJK-LMNPQ-RSTUV (~100 bits). */
    fun generateRecoveryKey(): String {
        val alphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789" // no ambiguous 0/O/1/I
        val bytes = ByteArray(20).also { SecureRandom().nextBytes(it) }
        val sb = StringBuilder()
        for (i in bytes.indices) {
            sb.append(alphabet[(bytes[i].toInt() and 0xFF) % alphabet.length])
            if ((i + 1) % 5 == 0 && i < bytes.size - 1) sb.append('-')
        }
        return sb.toString()
    }

    fun randomSalt(): ByteArray = ByteArray(SALT_LEN).also { SecureRandom().nextBytes(it) }

    /** wrapKey (stays on device) + authKey (proves possession to the server). */
    fun deriveRecovery(recoveryKey: String, salt: ByteArray): RecoveryKeys {
        val normalized = recoveryKey.replace(Regex("[\\s-]"), "").uppercase()
        val spec = PBEKeySpec(normalized.toCharArray(), salt, ITERATIONS, KEY_BITS)
        val master = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
        return RecoveryKeys(hmacSha256(master, "recovery-wrap"), hmacSha256(master, "recovery-auth").toHex())
    }

    /** AES-GCM wrap → base64(iv || ciphertext+tag). */
    fun wrapBytes(plain: ByteArray, wrapKey: ByteArray): String {
        val iv = ByteArray(IV_LEN).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(wrapKey, "AES"), GCMParameterSpec(TAG_BITS, iv))
        return b64(iv + cipher.doFinal(plain))
    }

    /** Reverses [wrapBytes]; throws if the wrap key is wrong. */
    fun unwrapBytes(blobB64: String, wrapKey: ByteArray): ByteArray {
        val combined = unb64(blobB64)
        val iv = combined.copyOfRange(0, IV_LEN)
        val ct = combined.copyOfRange(IV_LEN, combined.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(wrapKey, "AES"), GCMParameterSpec(TAG_BITS, iv))
        return cipher.doFinal(ct)
    }

    /** Public base64 helper for callers needing to send raw bytes (e.g. the salt). */
    fun toBase64(bytes: ByteArray): String = b64(bytes)

    // ---- primitives ----

    private fun hmacSha256(key: ByteArray, message: String): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(message.toByteArray(Charsets.UTF_8))
    }

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
    private fun b64(bytes: ByteArray) = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun unb64(s: String) = Base64.decode(s, Base64.NO_WRAP)
}

File: companion-ios/SentryKey/Cloud/CloudCrypto.swift
import Foundation
import CryptoKit
import CommonCrypto

/// Zero-knowledge cloud-account crypto. Byte-for-byte compatible with the web
/// dashboard (`crypto.js`) and Android (`CloudCrypto.kt`), pinned to a shared
/// vector (username "alice" / "supersecuremasterpassword"):
///   authKey = 68c6d3429e6d6b8025aa38fc281779b1cf308a90e6422888c94aceeb621bc0f4
///   encKey  = 3294dbf0fca8ca2c526b1397fb9bd6c1175caf5674fb5ccbf99af4956a9dbb03
///
///   masterKey = PBKDF2-HMAC-SHA256(password, salt = lowercased username, 210k, 32B)
///   authKey   = HMAC-SHA256(masterKey, "auth-key")        -> hex, sent to server
///   encKey    = HMAC-SHA256(masterKey, "encryption-key")  -> 32B, never leaves device
/// Vault is sealed with AES-256-GCM using encKey directly (envelope salt is decorative).
enum CloudCrypto {
    static let iterations = 210_000

    struct UserKeys {
        let authKey: String
        let encKey: Data
    }

    static func deriveUserKeys(username: String, password: String) -> UserKeys {
        let cleanUser = username.trimmingCharacters(in: .whitespaces).lowercased()
        let master = pbkdf2SHA256(password: password, salt: Data(cleanUser.utf8), iterations: iterations, keyLen: 32)
        let key = SymmetricKey(data: master)
        let authMac = HMAC<SHA256>.authenticationCode(for: Data("auth-key".utf8), using: key)
        let encMac = HMAC<SHA256>.authenticationCode(for: Data("encryption-key".utf8), using: key)
        let authHex = Data(authMac).map { String(format: "%02x", $0) }.joined()
        return UserKeys(authKey: authHex, encKey: Data(encMac))
    }

    // MARK: - Account recovery (matches web crypto.js / Android CloudCrypto.kt)

    struct RecoveryKeys {
        let wrapKey: Data
        let authKey: String
    }

    /// A readable one-time recovery key, e.g. ABCDE-FGHJK-LMNPQ-RSTUV (~100 bits).
    static func generateRecoveryKey() -> String {
        let alphabet = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789") // no ambiguous 0/O/1/I
        var bytes = Data(count: 20)
        _ = bytes.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 20, $0.baseAddress!) }
        var out = ""
        for (i, b) in bytes.enumerated() {
            out.append(alphabet[Int(b) % alphabet.count])
            if (i + 1) % 5 == 0 && i < bytes.count - 1 { out.append("-") }
        }
        return out
    }

    static func randomSalt() -> Data {
        var s = Data(count: 16)
        _ = s.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 16, $0.baseAddress!) }
        return s
    }

    /// wrapKey (stays on device) + authKey (proves possession to the server).
    static func deriveRecovery(recoveryKey: String, salt: Data) -> RecoveryKeys {
        let norm = recoveryKey
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "-", with: "")
            .uppercased()
        let master = pbkdf2SHA256(password: norm, salt: salt, iterations: iterations, keyLen: 32)
        let key = SymmetricKey(data: master)
        let wrap = HMAC<SHA256>.authenticationCode(for: Data("recovery-wrap".utf8), using: key)
        let auth = HMAC<SHA256>.authenticationCode(for: Data("recovery-auth".utf8), using: key)
        let authHex = Data(auth).map { String(format: "%02x", $0) }.joined()
        return RecoveryKeys(wrapKey: Data(wrap), authKey: authHex)
    }

    /// AES-GCM wrap → base64(iv(12) || ciphertext || tag).
    static func wrapBytes(_ plain: Data, wrapKey: Data) -> String {
        var iv = Data(count: 12)
        _ = iv.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 12, $0.baseAddress!) }
        let sealed = try! AES.GCM.seal(plain, using: SymmetricKey(data: wrapKey), nonce: try! AES.GCM.Nonce(data: iv))
        return (iv + sealed.ciphertext + sealed.tag).base64EncodedString()
    }

    /// Reverses wrapBytes; nil if the wrap key is wrong.
    static func unwrapBytes(_ blobB64: String, wrapKey: Data) -> Data? {
        guard let combined = Data(base64Encoded: blobB64), combined.count > 12 + 16 else { return nil }
        let iv = combined.prefix(12)
        let body = combined.dropFirst(12)
        let ct = body.prefix(body.count - 16)
        let tag = body.suffix(16)
        guard let box = try? AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: iv), ciphertext: ct, tag: tag),
              let plain = try? AES.GCM.open(box, using: SymmetricKey(data: wrapKey)) else { return nil }
        return plain
    }

    static func encryptWithKey(_ plaintext: String, encKey: Data) -> String {
        var salt = Data(count: 16)
        var iv = Data(count: 12)
        _ = salt.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 16, $0.baseAddress!) }
        _ = iv.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 12, $0.baseAddress!) }

        let key = SymmetricKey(data: encKey)
        let sealed = try! AES.GCM.seal(Data(plaintext.utf8), using: key, nonce: try! AES.GCM.Nonce(data: iv))
        let combined = sealed.ciphertext + sealed.tag   // ciphertext || tag, matches web/Android

        let env: [String: Any] = [
            "app": "SentryKey",
            "version": 1,
            "encrypted": true,
            "kdf": "PBKDF2WithHmacSHA256",
            "iterations": iterations,
            "salt": salt.base64EncodedString(),
            "iv": iv.base64EncodedString(),
            "ciphertext": combined.base64EncodedString()
        ]
        let data = try! JSONSerialization.data(withJSONObject: env, options: [.prettyPrinted])
        return String(decoding: data, as: UTF8.self)
    }

    static func decryptWithKey(_ envelopeJson: String, encKey: Data) -> String? {
        guard let data = envelopeJson.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              (obj["encrypted"] as? Bool) == true,
              let ivB64 = obj["iv"] as? String,
              let ctB64 = obj["ciphertext"] as? String,
              let iv = Data(base64Encoded: ivB64),
              let combined = Data(base64Encoded: ctB64),
              combined.count > 16 else { return nil }

        let ciphertext = combined.prefix(combined.count - 16)
        let tag = combined.suffix(16)
        let key = SymmetricKey(data: encKey)
        guard let box = try? AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: iv), ciphertext: ciphertext, tag: tag),
              let plain = try? AES.GCM.open(box, using: key) else { return nil }
        return String(decoding: plain, as: UTF8.self)
    }

    // MARK: - PBKDF2 (CommonCrypto)

    private static func pbkdf2SHA256(password: String, salt: Data, iterations: Int, keyLen: Int) -> Data {
        var derived = Data(count: keyLen)
        let pw = Data(password.utf8)
        _ = derived.withUnsafeMutableBytes { dp in
            salt.withUnsafeBytes { sp in
                pw.withUnsafeBytes { pp in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        pp.bindMemory(to: Int8.self).baseAddress, pw.count,
                        sp.bindMemory(to: UInt8.self).baseAddress, salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(iterations),
                        dp.bindMemory(to: UInt8.self).baseAddress, keyLen
                    )
                }
            }
        }
        return derived
    }
}
